### ISSUE TITLE
Ability to do your own each for body

### ISSUE THREAD

**Issue by cah-briangantzler** [miguelcobain/ember-yeti-table#170]
What do you think about a switch, or a different component similar to body but would only yield the data and the row component, leaving the each to be down by the block.

We were experimenting with ember-animated to get deleted rows to fade, and it doesnt look like animated-if will work. It seems to work only on single items, not repeating groups.

I think we need access to replace the each with an animated-each.

Would be easy, can throw a PR together. Any preference on a switch where body yields something different (seems would be confusing) or in addition to table yielding body it also yields bodyOnly (terrible name, cant think of a good one at the moment)

**Comment by miguelcobain** [comment]
I think this is a good idea.
It would open up other possibilities such as rendering custom specific rows that do not correspond to any data or even using [vertical-collection](https://github.com/html-next/vertical-collection).

Probably the easiest way to do this would be to do an `{{#if` on some argument here, in the body component:
https://github.com/miguelcobain/ember-yeti-table/blob/master/addon/components/yeti-table/body/template.hbs#L3-L8

And just yield the whole data with the row component, like you mentioned.
Can't think of a good name for this argument, though. :/

`@unroll={{false}}`? `@each={{false}}`?
Also considering if this should just be the default behaviour of the body component when you use a block.

Basically

```hbs
<table.body/>
```

would do the each and

```hbs
<table.body>
  ...
</table.body>
```

wouldn't do an each.

This would of course require a major bump as it is a breaking change.

**Comment by cah-briangantzler** [comment]
Thats why I thought maybe yielding a different body component would allow a transition why still keeping the old behaviour. Very similar to head and header. Dont ahve a good name for an alt body component or the switch. Will think on it.

**Comment by miguelcobain** [comment]
If the `<table.head>` component was called `<table.thead>` (as in, just render the `<thead>` element, we could have a `<table.body>` (the one we have now) and a new `<table.tbody>` (as in, just render the `<tbody>` element).

To avoid a major bump, we could yield both the `table.head` and `table.thead` for now, and deprecate the usage of the `table.head` component.

This scenario seems consistent to me.

**Comment by cah-briangantzler** [comment]
I like it

**Comment by miguelcobain** [comment]
Great!

The plan I'm thinking for this would be:
- I don't think we need to create a new body component that is almost a copy of the existing one without the each. We could make the current body component do that "if" on a private argument like `@each={{false}}` and then on the `YetiTable` component we yield the body component twice, but one of them named as `tbody` and with that internal `@each={{false}}` argument
- for the new `thead` component:
  - rename the internal `yeti-table/head` to `yeti-table/thead` for consistency
  - yield that component twice, one as `head` and another as `thead`. The head one would have an internal `@shouldDeprecate={{true}}` argument
  - inside the thead component's constructor we would `if (this.shouldDeprecate)` and deprecate using Ember's [deprecate](https://api.emberjs.com/ember/3.16/functions/@ember%2Fdebug/deprecate) function
- documentation would have to be updated

**Comment by cah-briangantzler** [comment]
That would leave `body` and `tbody` maintaining both. Use `body` when you want an `each` use `tbody` when you dont and ..... either with no block because they are both the same, that seems misleading.

How about deprecate `body` same as head, make the param not private and the default to be not to do the each (If your serious about making that the new default) and let user set `@doEach` (still not a great name) to true if they want the each done for them

The migration path would be to change to `tbody` and set `@doEach` to true. If they have to touch the line already to change to `tbody` not that much of an ask to add the `@doEach`. ONLY Providing you really want that to be the new default

**Comment by miguelcobain** [comment]
I don't understand how it is misleading.
`table.body` would be exactly the same as we have now and `table.tbody` wouldn't do the each.
The fact that both use the same component under the hood is just an implementation detail. Users would never see that.

You kind of convinced me of not using any `@doEach`-like argument on *user land* when you mentioned that we have head and header components. Doing the same for the body would be consistent.

No need to have the user make the `{{#each` by default when we have a `table.body` and a `table.tbody`. There is no default, the user just chooses what variant of the body he wants, just like the header.

The renaming of `head` to `thead` is just for consistency with `tbody`.

**Comment by cah-briangantzler** [comment]
So you are at two seperate components, `tbody` and `body`?

Where i said misleading was more of, if you are using the none-block version, which do you use? since `<table.body />` and <table.tbody />` would do the same thing. Which are we suggesting they use? And we should suggest one as if you choose to deprecate one in the future, you want it to be the one we didnt suggest using.

Agree on `head` rename

**Comment by miguelcobain** [comment]
> since <table.body /> and <table.tbody /> would do the same thing. Which are we suggesting they use?

No, `<table.body/>` would do the each (with or without a block) and `<table.tbody/>` would **never** do the each.

An example without an each-ing body would be:

```hbs
<YetiTable @data={{this.data}} as |table|>

  <table.header as |header|>
    <header.column>
      First name
    </header.column>
    <header.column>
      Last name
    </header.column>
    <header.column>
      Points
    </header.column>
  </table.header>

  <table.tbody as |body rowsData|>
    {{#each rowsData as |person|}}
      <body.row as |row|>
        <row.cell>
          {{person.firstName}}
        </row.cell>
        <row.cell>
          {{person.lastName}}
        </row.cell>
        <row.cell>
          {{person.points}}
        </row.cell>
      </body.row>
    {{/each}}
  </table.tbody>
</YetiTable>
```

**Comment by cah-briangantzler** [comment]
So `tbody` does not support a non-block format.

I agree

**Comment by miguelcobain** [comment]
> So tbody does not support a non-block format.

Yes, precisely.
I was just suggesting a way to implement `tbody` without creating a new component (reusing the current body component).

Basically, the main yeti-table template would yield something like this:

```hbs
{{yield (hash
  ...

  body=(component "yeti-table/body" data=this.processedData columns=this.columns theme=this.mergedTheme parent=this)
  tbody=(component "yeti-table/body" data=this.processedData columns=this.columns theme=this.mergedTheme parent=this _noEach=true)

  ...
)}}
```

And inside the body template we would

```hbs
{{#if @_noEach}}
  {{!-- yield the whole data array plus the body hash with the row component --}}
{{else}}
  {{!-- do the each, basically what we have now --}}
{{/if}}
```

**Comment by cah-briangantzler** [comment]
The non-block needs to be in there somewhere ... so 

```hbs
{{#if @_noEach}}
  {{yield
      (hash row=(component "yeti-table/body/row" theme=@theme onClick=this.onRowClick columns=@columns))
      rowData
  }}
{{else}}  
  {{#if hasBlock}}
  
    {{#each @data as |rowData index|}}
      {{yield
        (hash row=(component "yeti-table/body/row" theme=@theme onClick=this.onRowClick columns=@columns))
        rowData
        index}}
    {{/each}}
  
  {{else}}
  
    {{#each @data as |rowData|}}
      {{#yeti-table/body/row
        theme=@theme
        onClick=(action this.onRowClick rowData) columns=@columns as |row|}}
  
        {{#each @columns as |column|}}
          {{#row.cell class=column.columnClass}}
            {{#if column.prop}}
              {{get rowData column.prop}}
            {{else}}
              {{rowData}}
            {{/if}}
          {{/row.cell}}
        {{/each}}
  
      {{/yeti-table/body/row}}
    {{/each}}
  {{/if}}
{{/if}}```

**Comment by cah-briangantzler** [comment]
I think we have a plan

Looks like we were already using tbody in the theme 
```js
@reads('theme.tbody')
```

**Comment by miguelcobain** [comment]
Hmm, you would never use `tbody` without a block, right?

If you did, maybe the default body behavior would be a good default?

That would make it:

```hbs
{{#if hasBlock}}
  
  {{#if @_noEach}}
    {{yield
      (hash row=(component "yeti-table/body/row" theme=@theme onClick=this.onRowClick columns=@columns))
      @data}}
  {{else}}
    {{#each @data as |rowData index|}}
      {{yield
        (hash row=(component "yeti-table/body/row" theme=@theme onClick=this.onRowClick columns=@columns))
        rowData
        index}}
    {{/each}}
  {{/if}}

{{else}}

  {{#each @data as |rowData|}}
    <YetiTable::Body::Row
      @theme={{@theme}}
      @onClick={{fn this.onRowClick rowData}} @columns={{@columns}} as |row|>

      {{#each @columns as |column|}}
        <row.cell @class={{column.columnClass}}>
          {{#if column.prop}}
            {{get rowData column.prop}}
          {{else}}
            {{rowData}}
          {{/if}}
        </row.cell>
      {{/each}}

    </YetiTable::Body::Row>
  {{/each}}
{{/if}}
```

**Comment by miguelcobain** [comment]
Well, maybe you could want an empty table for some reason, but you could also pass in an empty block for that.

**Comment by cah-briangantzler** [comment]
Yea, just if no block produces the same thing for each component, we should encourage one over the other. Just as guidance and pick the one least likely to be deprecated in the future. If that can be predicted.

**Comment by miguelcobain** [comment]
Ah, I see what you mean.

We basically have 3 cases for body:
1. batteries included™ `<table.body/>`
2. the each-like `<table.body>...</table.body>`
3. the full-customization `<table.tbody>...</table.tbody>` where you have to do the each your self

If we want to keep all 3, then we must have the 2 components.
The only alternative would be to support 1 and 3, in which case a single body component would be enough, is that right?
The idea of having a single body component is very compelling to me. However that would require a major bump, right?
What is your suggestion?

**Comment by cah-briangantzler** [comment]
No I think you want to support all three, if people dont want to do the each, I dont see a problem with helping them out.
But I do vote for two components, one that supports case 1 and 2 and one that supports case 3 only. Just to avoid the "Which one do I use?" question



FYI did you also know that the below works, and you dont have to re-export in the app dir?
```
  {{component "ember-yeti-table@yeti-table/body/row"}}
```

```
  <EmberYetiTable@YetiTable />
```

I dont know how long this has been working.

I saw them doing this here https://github.com/adopted-ember-addons/ember-sortable/blob/master/addon/templates/components/sortable-group.hbs and the component version supports quite a ways back

**Comment by miguelcobain** [comment]
wat? I didn't know that.
They seem to be re-exporting the component there: https://github.com/adopted-ember-addons/ember-sortable/blob/master/app/components/sortable-item.js

**Comment by cah-briangantzler** [comment]
If you use that syntax internally, you only would re-export yeti-table in the app directory, and wouldnt have to do all the other internal ones. That does mean the end user though cant redefine or extend say row. Not sure if you wanted them to.

Yea, I guess they are re-exporting it in app. But I tried it in a few of my internal addons and did not re-export and worked fine

**Comment by cah-briangantzler** [comment]
I would assume the re-export is from the ember generator and they just didnt delete it. But I tried and they are not needed

**Comment by miguelcobain** [comment]
well, if it works without them, feel free to update them, please.
They are supposed to be private components.

We currently support latest 3.4.X, latest 3.5.X and 3.6+, per the docs.

**Comment by cah-briangantzler** [comment]
The only thing is, I dont see this documented anywhere.

We should ask someone in the discord channels if this is a public API they are supporting, or someone just found out if works but shouldnt be using it.

**Comment by cah-briangantzler** [comment]
And if someone out there really did extend one of the private ones, do you consider that a breaking change, or it was private and they shouldnt have done that (It was exported in the app folder. is that private?)

**Comment by miguelcobain** [comment]
We don't really say it is private anywhere. But I can't think of a reason why anybody would do that. Also, the new component the user extends wouldn't really be usable.

The main yeti table keeps passing in stuff internally when it yields contextual components. If you create your own, you wouldn't be able to do that, I think.

Yeti table is completely customisable in how it renders its markup (especially after this tbody update), so if users are doing it its probably because of a code smell.

I wouldn't worry about it.

**Comment by cah-briangantzler** [comment]
I have overridden a few addons internal components from time to time :) And felt the burden of having to maintain it. But that was before I just started making PRs to the addons themselves.

Im sure there are others out there doing the same. Maybe put a something in the docs to let them know if they upgrade and they did that, it wont work anymore.

**Comment by miguelcobain** [comment]
The changelog would be a good place to document that.

**Comment by cah-briangantzler** [comment]
In the middle of doing this...... Should foot also be renamed to tfoot? and deprecate foot?

**Comment by miguelcobain** [comment]
That's a good point. I'd say yes because it makes names more consistent.

**Comment by cah-briangantzler** [comment]
Work and illness have gotten in the way, sorry, Will get to this when I can. On the plus side I rwjblue confirmed today the `<addon@component>` is supported, so will work that into making all the components private etc yeti-table proper

**Comment by miguelcobain** [comment]
@cah-briangantzler perfect!
In the meantime, I added some more features and components. Check the changelog. Shouldn't change any part of the plan we laid out.

**Comment by cah-briangantzler** [comment]
FYI. Since you are doing more of a PODS layout, when I do the component as private I have to do `row=(component "ember-yeti-table@yeti-table/body/row/component" `. Note the component on the end. Hope that isnt a problem

**Comment by miguelcobain** [comment]
If it works, I don't think it makes a difference. 👍

**Comment by miguelcobain** [comment]
Closed by #192
